[LLVM] Slay undead copysign code

workingjubilee opened the pull request:
Frontends that emit LLVMIR would like to be able to count on float operations that can be encoded as simple bit-operations as lowering to machine-code routines embedded into the binary. This is desired so that they do not affect what one might think of as "freestanding" binaries. SelectionDAG[^1] and GlobalISel[^2] both explicitly handle this lowering for copysign.

However, later code implies that it is, at least in theory, possible to lower to runtime libcalls, as indicated by its entry in llvm/include/llvm/IR/RuntimeLibcalls.def[^3]! Yet this code is largely impotent. We seem to need to preserve runtime libcalls for architecture-specific variants of these float operations that have architecture-specific semantics. However, ordinary IEEE754 floats have nonesuch: these are bitops.

So, purge this dead code, that it may plague us no more with its revenant half-life.

[^1]:  https://github.com/llvm/llvm-project/blob/46944b0cbc9a9d8daad0182c40fcd3560bc9ca35/llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp#L7772-L7811
[^2]: https://github.com/llvm/llvm-project/blob/75103aae4a9d22e2c46068f2160f2dddd6ad2116/llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp#L1646-L1707
[^3]: https://github.com/llvm/llvm-project/blob/46944b0cbc9a9d8daad0182c40fcd3560bc9ca35/llvm/include/llvm/IR/RuntimeLibcalls.def#L287-L291

workingjubilee committed:
[SelectionDAG] Only lower COPYSIGN_PPC_F128 to copysignl

This reduces the burden on frontends that wish to support float ops
without needing a C compiler to build LLVM's compiler-rt for that target,
e.g. so that they can be a fully self-contained toolchain for bare-metal.

All other floats are expanded for all current architectures just fine.
PowerPC, however, does not efficiently legalize its very own float.

topperc commented on llvm/lib/CodeGen/IntrinsicLowering.cpp:
```diff
@@ -58,7 +58,9 @@ static Value *LowerBSWAP(LLVMContext &Context, Value *V, Instruction *IP) {
   switch(BitSize) {
-  default: llvm_unreachable("Unhandled type size of value to byteswap!");
+  default:
+    assert(false && "Unhandled type size of value to byteswap!");
```
Seems unrelated?

topperc commented on llvm/lib/CodeGen/IntrinsicLowering.cpp:
```diff
@@ -203,7 +205,9 @@ static void ReplaceFPIntrinsicWithCall(CallInst *CI, const char *Fname,
   switch (CI->getArgOperand(0)->getType()->getTypeID()) {
-  default: llvm_unreachable("Invalid type in intrinsic");
+  default:
+    assert(false && "Invalid type in intrinsic");
```
Seems unrelated?

topperc commented on llvm/lib/CodeGen/IntrinsicLowering.cpp:
```diff
@@ -438,7 +442,15 @@ void IntrinsicLowering::LowerIntrinsicCall(CallInst *CI) {
-    ReplaceFPIntrinsicWithCall(CI, "copysignf", "copysign", "copysignl");
+    switch (CI->getArgOperand(0)->getType()->getTypeID()) {
+    default:
+      assert(false && "only need a copysign libcall for arch-specific floats");
```
Why assert instead of llvm_unreachable?

workingjubilee replied:
Because its name is usually false[^1][^2][^3][^4][^5][^6][^7], and if LLVM is compiled with `LLVM_UNREACHABLE_OPTIMIZE=ON`, then:
> When ON (default) then llvm_unreachable() is considered "undefined behavior" and optimized as such. When OFF it is instead replaced with a guaranteed "trap".

I am not great at reading C++, but the code here does not seem to have a type-checked guarantee enforcing that this is _actually_ unreachable. It is relying on callers not calling it in a mistaken way. The required invariant is not even documented as causing UB if violated, as far as I can tell: https://github.com/llvm/llvm-project/blob/6de5305b3d7a4a19a29b35d481a8090e2a6d3a7e/llvm/include/llvm/CodeGen/IntrinsicLowering.h#L27-L42

Is there such a guarantee that I am not seeing?

[^1]: https://github.com/llvm/llvm-project/issues/62012
[^2]: https://github.com/llvm/llvm-project/issues/104718
[^3]: https://github.com/llvm/llvm-project/issues/62252
[^4]: https://github.com/llvm/llvm-project/issues/59017
[^5]: https://github.com/llvm/llvm-project/issues/109251
[^6]: https://github.com/llvm/llvm-project/issues/48149
[^7]: https://github.com/llvm/llvm-project/issues/85783

topperc replied:
Is the `break;` on a release build any better? It will exit the switch and hit the `CI->eraseFromParent() ` which will likely leave the IR in a broken state.

Maybe `report_fatal_error` would be more appropriate?

workingjubilee replied:
`report_fatal_error` sounds good to me. I just hadn't thought terribly hard about error-handling aside from "hm, this change might let weird IR flow downstream, maybe this spot which serves as a backstop for such should assert more".

workingjubilee replied:
My hypothesis is still that it makes a 0% chance 100 times more likely... so, 0%... but if I am wrong, well, best to discover that quickly.

arsenm commented:
Regardless of the usage, I think these should still be recorded in RuntimeLibcalls. Ideally it would be a complete database of all calls for the system, not just used

workingjubilee commented:
> Regardless of the usage, I think these should still be recorded in RuntimeLibcalls. Ideally it would be a complete database of all calls for the system, not just used

Of all calls for... what? All function calls that are hypothetically possible at runtime? Used for what purpose? Based on what conditions? What are we assuming is linked and therefore imposing on frontends that they must guarantee is linked?

Can we have some way of marking the ones that aren't actually going to, and should not, get spontaneously lowered to?

RalfJung commented:
> Regardless of the usage, I think these should still be recorded in RuntimeLibcalls. Ideally it would be a complete database of all calls for the system, not just used

It would be good to know which functions LLVM might actually use so that frontends have a chance of ensuring they are all provided. And it would be good for that list to be as short as possible so that frontends don't spend effort trying to provide some functions that LLVM will not actually use.

workingjubilee commented:
If a true database of all function calls that _might_ be made at runtime is desired, @arsenm, and not "all calls *that LLVM might generate*", then the .def file in question is hopelessly short. You have approximately the entirety of libc (minus _some_ of math.h), Win32, and POSIX APIs to include. I don't think setting _this_ list back by even a few more functions than this would be appreciable.

It doesn't even have:
- abs
- fdim
- nan
- remainder
- remquo
- hypot
- atan2
- acosh
- asinh
- atanh
- erf
- erfc
- lgamma
- tgamma
- modf (not to be confused with fmodf)
- scalbn
- log1p
- logb
- ilogb
- expm1
- nextafter
- nexttoward

...or all their differently-sized variants.

workingjubilee commented:
@topperc This is using `report_fatal_error` now.

Also, good news: I seem to have found the spot where SelectionDAG expands copysign that catches this before we ever hit the libcall legalization: https://github.com/llvm/llvm-project/blob/75103aae4a9d22e2c46068f2160f2dddd6ad2116/llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp#L1646-L1707

workingjubilee commented:
And as I suspected, trying to simplify the code further hits the PowerPC float case on some of the regression tests for that target, so this is exactly the correct scope.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:
```diff
@@ -1649,12 +1649,13 @@ void DAGTypeLegalizer::ExpandFloatRes_FCEIL(SDNode *N,
-                                        RTLIB::COPYSIGN_F128,
-                                        RTLIB::COPYSIGN_PPCF128), Lo, Hi);
+
+  auto VT = N->getValueType(0);
```
No auto

workingjubilee replied:
Sure. I'm curious why clang-format doesn't lint on it then, tho'.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:
```diff
@@ -1649,12 +1649,13 @@ void DAGTypeLegalizer::ExpandFloatRes_FCEIL(SDNode *N,
-                                        RTLIB::COPYSIGN_F80,
-                                        RTLIB::COPYSIGN_F128,
-                                        RTLIB::COPYSIGN_PPCF128), Lo, Hi);
+
```
Did this never handle the case where the 2 operands have different types?

workingjubilee replied:
That does in fact seem to be the case, yes.

workingjubilee replied:
I'm not sure why it would handle that case, though? We're in LLVM IR. The floats should be of the same type. Instructions like `fadd` and such demand it. The language reference always uses identical types for this intrinsic, as presented here: https://llvm.org/docs/LangRef.html#llvm-copysign-intrinsic

That is presumably for a reason? If the frontend wants to do type promotion, that is a C-like language semantic that clang can apply. The response that LLVM should have to these being different types is `report_fatal_error`, isn't it?

workingjubilee replied:
And if it is indeed erroneous, it should have been handled already, I would think.

arsenm replied:
This is not in LLVM IR. The IR intrinsic requires both types to match (and I consider this a flaw in the intrinsic). ISD::FCOPYSIGN does not, and the [combiner forms them](https://github.com/llvm/llvm-project/blob/f0ed31ce4b63a5530fd1de875c0d1467d4d2c6ea/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L17871)

So it's not erroneous. The type of the second operand does not matter

workingjubilee replied:
I guess we arrive back at "this code doesn't have to worry about handling the operands being of different sizes" but the other way around, then.

workingjubilee commented:
@arsenm I see. If you can coach me on some of the finer points of making a `.td` (I have the basic gist, but there will be some questions), I'd be happy to assemble something such that the data covered in those two files is in a tablegen, in a followup. Then I will be able to make sure that it or another file is supplied with the necessary information that frontends... frontends that aren't clang, anyways... need. And then it will cover libm _properly_, instead of the current state of affairs.

arsenm commented on llvm/include/llvm/IR/RuntimeLibcalls.def:
```diff
@@ -284,10 +284,7 @@ HANDLE_LIBCALL(FLOOR_F64, "floor")
 HANDLE_LIBCALL(FLOOR_PPCF128, "floorl")
-HANDLE_LIBCALL(COPYSIGN_F32, "copysignf")
-HANDLE_LIBCALL(COPYSIGN_F64, "copysign")
 HANDLE_LIBCALL(COPYSIGN_F80, "copysignl")
```
Weird to leave the l case. What was the PPC issue? We should be able to expand x86_fp80, ppcf128, and fp128 just fine?

jhuber6 reviewed (commented):
I like the idea of getting rid of target libcalls, linking in runtime libraries in the same TU is still an unsolved problem. Right now the first step was to only keep symbols alive if the backend actually emits them. Next step would probably have the IR transform LLVM IR intrinsics to libcalls during IR linking and then mark them as `nobuiltin` so they don't get transformed again.

Seems weird that this doesn't handle the x86 / ppc cases. They're weird but it should just be copying a bit I would think.

workingjubilee commented:
I _believe_ the PowerPC target code has setOperationAction for ppc_f128 in a way that somehow still finds this codepath. As it is logically the only user of this codepath, it seemed harmless to just leave it. If you want, I can go muck about inside llvm/lib/Target/PowerPC/PPCISelLowering.cpp so that the regression tests pass and we can completely remove this, but my initial inclination was to not bother because the Rust frontend doesn't use these types at all and I will be hacking pretty wildly at that point.

I believe the x87 case might actually be handled... "correctly". At least in the sense that the code produced by LLVM should be no more incorrect than it was before, for this target.

arsenm commented:
> I _believe_ the PowerPC target code has setOperationAction for ppc_f128 in a way that somehow still finds this codepath. 

It actually hits the default expand handling. ppc_fp128 surprisingly isn't a legal type:

```
Legalizing node: t20: ppcf128 = fcopysign t5, t4
Analyzing result type: ppcf128
Expand float result: t20: ppcf128 = fcopysign t5, t4
Creating new node: t23: i128 = bitcast t5
Creating new node: t24: i64 = extract_element t23, Constant:i64<1>
Creating new node: t25: i64 = extract_element t23, Constant:i64<0>
Creating new node: t26: f64 = bitcast t25
Creating new node: t27: f64 = bitcast t24

```

and it does result in the call:

```
	lfs 2, .LCPI0_0@toc@l(3)
	bl copysignl
```

workingjubilee commented:
lmao PowerPC doesn't treat its own float as a legal type

workingjubilee commented:
I checked and the x87 float seems fine if this codepath is pruned. No visible regressions in LLVM's test suite. I mean, clang-format whines about my formatting again, but

arsenm reviewed (approved):
I guess this is fine, although unfortunate to leave ppc_fp128 behind.

workingjubilee commented:
...is getOperand actually getOperand**s**?

topperc commented:
> ...is getOperand actually getOperand**s**?

No. What makes you think that?

workingjubilee committed:
[PowerPC][SelectionDAG] Expand `@llvm.copysign.ppc_fp128` without copysignl

This allows ripping out the remaining copysignl infra.

workingjubilee commented:
@topperc Momentary delirium, it seems?

@arsenm re: your remarks about not leaving the PowerPC libcall behind, I figured out how to remove the code in question, after some wrangling, by doing the conversion in question. See the recently pushed commits.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:
```diff
@@ -1649,12 +1649,19 @@ void DAGTypeLegalizer::ExpandFloatRes_FCEIL(SDNode *N,
+
+  assert(N->getValueType(0) == MVT::ppcf128 &&
+         "Logic only correct for ppcf128!");
+  SDLoc DL = SDLoc(N);
+  SDValue Tmp = SDValue();
```
```suggestion
  SDLoc DL(N);
  SDValue Tmp;
```

arsenm commented on llvm/test/CodeGen/PowerPC/copysignl.ll:
```diff
@@ -71,12 +74,12 @@ entry:
-; CHECK: bl copysignl
-; CHECK: blr
+; CHECK-LABEL:     @foo_lf
+; CHECK-NOT:         bl copysignl
```
-NOT checks are fragile, use positive checks for what is emitted

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:
```diff
@@ -1649,12 +1649,19 @@ void DAGTypeLegalizer::ExpandFloatRes_FCEIL(SDNode *N,
+
+  Hi = DAG.getNode(ISD::FCOPYSIGN, DL, Tmp.getValueType(), Tmp,
+                   N->getOperand(1));
+  // a double-double is Hi + Lo, so if Hi flips sign, so must Lo
```
```suggestion
  // A double-double is Hi + Lo, so if Hi flips sign, so must Lo
```

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeFloatTypes.cpp:
```diff
@@ -1649,12 +1649,19 @@ void DAGTypeLegalizer::ExpandFloatRes_FCEIL(SDNode *N,
+  // a double-double is Hi + Lo, so if Hi flips sign, so must Lo
+  Lo = DAG.getSelectCC(DL, Tmp, Hi, Lo,
+                       DAG.getNode(ISD::FNEG, DL, Lo.getValueType(), Lo),
+                       ISD::SETEQ);
```
Using SETEQ for FP types is weird, since it's the case that may be oeq or ueq. Can this break the handling of nan sign bits?

Can this just be emitted as another copysign?

workingjubilee replied:
Hmm. Because the rule is Hi + Lo == "the real float", if one is NaN then logically they both are, aren't they?

Though, to be clear:
- If the double-double is NaN or Infinity, the Hi value must encode NaN or Infinity.
- If the double-double is Infinity, then Lo must not be NaN.
- Otherwise, the value of Lo is not significant.

workingjubilee replied:
Basically, the answer of "how does `ppc_fp128` work for this" is "mmmostly by not really caring too hard, honestly".

Which doesn't mean I am against using the SETUEQ version. If I understand how the `getSelectCC` API works it does _sound_ more correct?

arsenm replied:
Unlike a true floating point operation, the "not significant" bits of a nan shouldn't change. It would be better if we weren't flipping the bit of the insignificant half here (or for any of fneg/fabs/fcopysign)

jayfoad commented:
> Slay undead copysign code

Suggest writing this in plainer English, e.g. "Remove unused copysign code".

workingjubilee committed:
[PowerPC][SelectionDAG] Use SETUEQ in copysign expansion

RKSimon commented on llvm/test/CodeGen/PowerPC/copysignl.ll:
```diff
@@ -91,8 +92,9 @@ define ppc_fp128 @foo_ld(double %a, double %b) #0 {
-; CHECK-VSX-NEXT:    beqlr 0
+; CHECK-VSX-NEXT:    fcmpu 0, 0, 1
+; CHECK-VSX-NEXT:    cror 20, 2, 3
+; CHECK-VSX-NEXT:    bclr 12, 20, 0
```
@workingjubilee please can you rebase after 32aa782ea297b3e0ec090cf8fc0055d00c99d24b to more clearly see the diff?

workingjubilee replied:
I already did that in my prior commit...

workingjubilee replied:
@RKSimon If I rebase on main, which I have to now, that will just dump 1000 files of diff on it in the GH view. It will obliterate any ability to see anything. If you wanted me to insert a commit at the start, before any code changes, I could have simply done that.

workingjubilee replied:
I know how to shove the rebase several times to make it so that GitHub renders a minimal diff eventually anyways, but **please never do this to anyone again.**

workingjubilee replied:
If you have write access, you could just shove my branch and get the commit where you want it anyways.

arsenm replied:
> If I rebase on main, which I have to now, that will just dump 1000 files of diff on it in the GH view.

No it won't

workingjubilee replied:
If you look at the Files view, yes, but it removes the ability to compare between the rebase versions.

arsenm replied:
Yes, but it doesn't really matter. I care about that < 2% of the time and have never looked at the GitHub version comparison

workingjubilee replied:
I suppose.

LLVM's a verrry long rebuild on my machine though, I'm not using a 100-core monster.

arsenm commented on llvm/test/CodeGen/PowerPC/ctrloop-cpsgn.ll:
```diff
@@ -1,28 +0,0 @@
-; RUN: llc -verify-machineinstrs < %s -mcpu=ppc | FileCheck %s
```
Lost this test?

workingjubilee replied:
according to https://github.com/llvm/llvm-project/commit/1cf48ab811b341cf2ee3d7e9a9c97cf2d7b53f64
> Don't form PPC CTR-based loops around a copysignl call
copysign/copysignf never become function calls (because the SDAG expansion code
does not lower to the corresponding function call, but rather directly
implements the associated logic), but copysignl almost always is lowered into a
call to the requested libm functon (and, thus, might clobber CTR).

So it seemed like it was now irrelevant.

workingjubilee replied:
Inserted a test with the baseline at start for this so you can examine the diff for it, if you want.

workingjubilee commented on llvm/test/CodeGen/PowerPC/ctrloop-cpsgn.ll:
```diff
@@ -1,9 +1,84 @@
+; CHECK-NEXT:    lwz 4, 88(1)
+; CHECK-NEXT:    stw 4, 104(1)
+; CHECK-NEXT:    lfd 0, 104(1)
+; CHECK-NEXT:    mtctr 3
```
when I said "the test seems irrelevant", we now emit an MTCTR instruction here, but the copysignl call we were guarding against is nowhere to be found, sooo

workingjubilee committed:
[PowerPC] regenerate ctrloop-cpsgn.ll test

workingjubilee committed:
[CodeGen] Demand llvm.copysign.f{16,32,64,80,128} lowers without libcalls

This makes real what is already true:
Copysign does not ever need to lower to runtime libcalls!
Its operation should be possible to always implement via bitops.

workingjubilee committed:
[PowerPC] Erase now-irrelevant test

workingjubilee commented:
@arsenm I can do further edits to this if you want but I cannot merge this, so I am wondering what it would take? I am wondering if it was a mistake to try to solve the problem you expressed discontent about. You seemed inclined to accept the PR in its then-current state but all successive edits that try to remove the vestigial code so that PowerPC would no longer be the odd one out that is "left behind" by this change have provoked many questions.
